tf2onnx/onnx_opset/nn.py

```diff
@@ -959,6 +959,92 @@ def version_11(cls, ctx, node, **kwargs):
 
 @tf_op(["FusedBatchNorm", "FusedBatchNormV2", "FusedBatchNormV3"])
 class BatchNorm:
+    @classmethod
+    def version_17(cls, ctx, node, **kwargs):
+        tf_type = node.type
+        input_rank = len(ctx.get_shape(node.input[0]))
+        if input_rank == 4:
+            spatial = 2
+        elif input_rank == 5:
+            spatial = 3
+        else:
+            raise ValueError("node input must be 4 or 5-dimensional, is {} now".format(input_rank))
+
+        node.type = "BatchNormalization"
+        # tf inputs: x, scale, bias, mean, variance
+        # tf outputs: y, batch_mean, batch_var
+        # a: data_format, epsilon, is_training
+        # onnx inputs: X, scale, B, mean, variance, attributes: epsilon, momentum=0.9, spatial : 1
+        # output: y, mean, var, savedmean, savedvar,
+        # detach unused outputs. While we could let the unused outputs dangle,
+        # some runtimes like pytorch/caffe2 do complain about it.
+
+        # onnx batchnorm requires same T for all inputs
+        mean_type = ctx.get_dtype(node.input[3])
+        x_dtype = ctx.get_dtype(node.input[0])
+        if x_dtype != mean_type:
+            # TODO: this works but more efficient would be to flip the other inputs. We'd need to check
+            # TODO: first if this works with the onnx implementation so its a later for now
+            ctx.insert_new_node_on_input(node, "Cast", node.input[0], to=mean_type)
+            # casting the input[0] will change the output dtype of bn so we need to cast back
+            cast_back_node = ctx.insert_new_node_on_output("Cast", node.output[0],
+                                                           name=utils.make_name(node.name) + "_castback",
+                                                           to=x_dtype)
+            ctx.set_dtype(cast_back_node.output[0], x_dtype)
+            ctx.copy_shape(node.name, cast_back_node.output[0])
+            ctx.set_dtype(node.output[0], mean_type)
+
+        consumers = [ctx.find_output_consumers(output_name) for output_name in node.output[1:]]
+        if not any(consumers):
+            new_output = [node.output[0]]
+            # the setter makes a copy of new_output
+            node.output = new_output
+
+        conv_convert_inputs(ctx, node, with_kernel=False, spatial=spatial)
+
+        inp_shape = ctx.get_shape(node.input[0])
+        inp_rank = len(inp_shape) if inp_shape is not None else None
+        scale_shape = ctx.get_shape(node.input[1])
+        mean_shape = ctx.get_shape(node.input[3])
+        var_shape = ctx.get_shape(node.input[4])
+        val_type = utils.map_onnx_to_numpy_type(ctx.get_dtype(node.input[1]))
+        is_training = node.get_attr_value('is_training', True)
+
+        if is_training and node.get_attr_value('exponential_avg_factor', 1.0) == 1.0:
```

Comment:
Thanks for your contributions!

Could you please create a common method where this loop could be refactored like [here](https://github.com/onnx/tensorflow-onnx/blob/b57fe915b3728fbf996b490170a374821dd8383a/tf2onnx/onnx_opset/nn.py#L835C9-L835C9)? So those different versions could call this common method and avoid duplicated code as much as possible.

Reply:
Ah I didn't know that. Done.

Reply:
Besides the pylint issue, looks like this change has broken some legacy tests, please also fix them. Thanks.

Reply:
Thanks! I mainly use ONNX for TRT, and don't know much about ONNX runtime. But it seems the ONNX layernorm doesn't support automatic broadcasting on scale tensor. So I need to do it manually. But I realized there is no broadcast op in ONNX either. Can you suggest how to broadcast an ONNX tensor?

Reply:
> Thanks! I mainly use ONNX for TRT, and don't know much about ONNX runtime. But it seems the ONNX layernorm doesn't support automatic broadcasting on scale tensor. So I need to do it manually. But I realized there is no broadcast op in ONNX either. Can you suggest how to broadcast an ONNX tensor?

Could you please follow the approach [here](https://github.com/onnx/tensorflow-onnx/blob/ca17b3c99eb37dc3f42c9be5a64620d247a24115/tf2onnx/onnx_opset/tensor.py#L4017C6-L4017C6) to try Expand op?